### PR TITLE
Fix performance in get_imports regexp

### PR DESCRIPTION
**Pull request by AlekseyLobanov**
# What does this PR do?

Just found that one of the Regular Expressions may be very slow (O(N^3)) on some special inputs.

Something like `'try:' + ' ' * 1500` may take a lot of time. I replaced it with the same logic but faster

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@gante

### CODE REVIEW

**Comment by AlekseyLobanov** [comment]
It is not a typo but still a small change with existing tests

File: src/transformers/dynamic_module_utils.py
Comment (Rocketknight1) [inline src/transformers/dynamic_module_utils.py:155]:
```suggestion
    content = re.sub(r"\s*try\s*:.*?except.*?:", "", content, flags=re.MULTILINE | re.DOTALL)
```

Comment (Rocketknight1) [reply src/transformers/dynamic_module_utils.py:155]: Looks good to me - can we take this even further and remove the whitespace matching here as well? With `re.DOTALL` they should be equivalent. Also, `re.MULTILINE` only affects [the handling of start and end characters](https://docs.python.org/3/library/re.html#re.MULTILINE), which aren't used here. Maybe we can drop that as well?

Comment (AlekseyLobanov) [reply src/transformers/dynamic_module_utils.py]: Yes, definitely. I removed  `re.MULTILINE` and extra `\s*`. Tests are fine.

**Comment by Rocketknight1** [comment]
This looks good now! It should be simpler and more performant, and shouldn't change which strings it matches.

I have one more suggestion, though. This regex looks like it would fail in cases when the strings "try" and "except" occur internally in some lines. This is probably very rare, but maybe we should change the regex to `\n\s+try` and `\n\s+except` to be secure against that?

**Comment by AlekseyLobanov** [comment]
>regex looks like it would fail in cases ...

Any regex should fail because even valid Python source is not defined by a regular language.
For example, code below also should be ignored, I believe.

```python
"""
try:
    pass
except:
    pass
"""
# code in multi line string literal
```

Should I add it in another PR with more tests? This PR is more about performance improvements.

Maybe we can wait for @gante to reply, because he wrote this lines.

**Review by Rocketknight1 — APPROVED** [review]
@AlekseyLobanov yes, good point - we can keep this PR simple. Want me to merge this one now, and if you think you can improve it further, you can open a new PR with added test(s)?

**Comment by gante** [comment]
> Maybe we can wait for @gante to reply, because he wrote this lines.

Hehe not true (https://github.com/huggingface/transformers/pull/23725 is the source of these lines), but the diff looks good to me :)

**Comment by Rocketknight1** [comment]
Yes, the diff looks good to me. Thanks for the PR!

**Pull request merged**
